vector/src/main/java/im/vector/app/features/onboarding/OnboardingViewModel.kt

```diff
@@ -588,93 +577,77 @@ class OnboardingViewModel @AssistedInject constructor(
-            getLoginFlow(homeServerConnectionConfig)
+            startAuthenticationFlow(homeServerConnectionConfig)
         }
     }
 
-    private fun getLoginFlow(homeServerConnectionConfig: HomeServerConnectionConfig,
-                             serverTypeOverride: ServerType? = null) {
+    private fun startAuthenticationFlow(homeServerConnectionConfig: HomeServerConnectionConfig, serverTypeOverride: ServerType? = null) {
         currentHomeServerConnectionConfig = homeServerConnectionConfig
 
         currentJob = viewModelScope.launch {
-            authenticationService.cancelPendingLoginOrRegistration()
-
-            setState {
-                copy(
-                        isLoading = true,
-                        // If user has entered https://matrix.org, ensure that server type is ServerType.MatrixOrg
-                        // It is also useful to set the value again in the case of a certificate error on matrix.org
-                        serverType = if (homeServerConnectionConfig.homeServerUri.toString() == matrixOrgUrl) {
-                            ServerType.MatrixOrg
-                        } else {
-                            serverTypeOverride ?: serverType
-                        }
-                )
-            }
-
-            val data = try {
-                authenticationService.getLoginFlow(homeServerConnectionConfig)
-            } catch (failure: Throwable) {
-                setState {
-                    copy(
-                            isLoading = false,
-                            // If we were trying to retrieve matrix.org login flow, also reset the serverType
-                            serverType = if (serverType == ServerType.MatrixOrg) ServerType.Unknown else serverType
-                    )
-                }
-                _viewEvents.post(OnboardingViewEvents.Failure(failure))
-                null
-            }
-
-            data ?: return@launch
-
-            // Valid Homeserver, add it to the history.
-            // Note: we add what the user has input, data.homeServerUrlBase can be different
-            rememberHomeServer(homeServerConnectionConfig.homeServerUri.toString())
+            setState { copy(isLoading = true) }
 
-            val loginMode = when {
-                // SSO login is taken first
-                data.supportedLoginTypes.contains(LoginFlowTypes.SSO) &&
-                        data.supportedLoginTypes.contains(LoginFlowTypes.PASSWORD) -> LoginMode.SsoAndPassword(data.ssoIdentityProviders)
-                data.supportedLoginTypes.contains(LoginFlowTypes.SSO)              -> LoginMode.Sso(data.ssoIdentityProviders)
-                data.supportedLoginTypes.contains(LoginFlowTypes.PASSWORD)         -> LoginMode.Password
-                else                                                               -> LoginMode.Unsupported
-            }
+            runCatching { startAuthenticationFlowUseCase.execute(homeServerConnectionConfig) }.fold(
```

Comment:
Using `withContext(Dispatchers.IO)` to `startAuthenticationFlowUseCase` will reduce the work on main thread. Or maybe using `viewModelScope.launch(Dispatchers.IO)` and exclude state updates to run on Main dispatcher

Reply:
I wasn't sure about this, the authentication service has a bunch of internal non thread safe state, we can rely on the lower layers (retrofit/awaitTransaction) for doing the dispatcher switching for us, at least whilst the other parts aren't thread safe

but it does mean there's some thread jumping happening~

Reply:
Yes, I was talking mainly for the logic that is running on the main thread, except retrofit calls etc that always run on another thread. But if we have thread safety issues, there are not much we can do!